campaignion_tracking/package/src/ga4.js

```diff
@@ -0,0 +1,426 @@
+   * @param {object} context the tracking context
+   */
+  callChangeHook (eventName, ga4Data, context) {
+    // maybe also inject tracker
+    if (typeof window.campaignion_tracking_change_msg === 'function') {
+      ga4Data = window.campaignion_tracking_change_msg('ga4', eventName, ga4Data, context)
+    }
+    return ga4Data
+  }
+
+  /**
+   * Maintains a context between event handling.
+   *
+   * You can provide context with a first event and call this method during
+   * handling.
+   * When another event arrives it can also enrich this context and read former
+   * context data as well.
+   *
+   * This context is saved in the browser storage to persist between page
+   * loads. It is reset on donationSuccess.
+   *
+   * @param {context} context Tracking context data.
+   */
+  updateContext (context = {}) {
+    // Check if context exists
+    if (context.donation) {
+      Object.assign(this._context.donation, context.donation)
+    }
+    if (context.node) {
+      Object.assign(this._context.node, context.node)
+    }
+    if (context.webform) {
+      Object.assign(this._context.webform, context.webform)
+    }
+    this.saveToStorage('context', this._context)
+  }
+
+  /**
+   * Handle "submission".
+   *
+   * Event data: { nid, sid, title }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_submission (eventName, eventData, context) {
+    this.printDebug('(handle)', eventName, eventData, context)
+    this.updateContext(context)
+
+    let submissionData = {
+      event: 'submission',
+      webform: {
+        nid: eventData.nid || null,
+        sid: eventData.sid || null,
+        title: eventData.title || null,
+      }
+    }
+    // Allow others to modify the data being sent to Google Analytics.
+    submissionData = this.callChangeHook(eventName, submissionData, this._context)
+    this.dataLayer.push(submissionData)
+    this.printDebug('(event)', submissionData)
+
+    if (eventData.optins.length > 0) {
+      let optinData = {
+        event: 'optin'
+      }
+      for (const channel of eventData.optins) {
+        optinData[channel.channel] = channel.value
+      }
+      // Allow others to modify the data being sent to Google Analytics.
+      optinData = this.callChangeHook(eventName, optinData, this._context)
+      this.dataLayer.push(optinData)
+      this.printDebug('(event)', optinData)
+    }
+  }
+
+  /**
+   * Handle "setDonationProduct".
+   *
+   * Remove a (donation) product when one was added before. Thus we have a
+   * "cart" with only 1 slot for 1 donation.
+   *
+   * Event data: { product, currencyCode }
+   *
+   * @param {String} eventName the event name
+   * @param {object} eventData data of the event
+   * @param {object} context context of the event
+   */
+  handle_setDonationProduct (eventName, eventData, context) {
```

Comment:
I think the logic of this function could be simplified if it would separate the product comparison from preparing the events. This also applies to the GA4 version of the next commit. So:

1. Build new product & compare it to the old, if equal then return.
2. Build events and send them.

Reply:
I’m afraid it's a bit more complicated. As much as I'd like to reduce unnecessary "setDonationProduct" events, this would need to keep track of the previous product data for comparison. Where/How would you suggest to do that? :thinking:

Reply:
Turns out the current version of the GA4 code didn’t even work :innocent: 

Have refactored the product comparison logic and deleting products, choosing session storage as this was already used by other bits of codes.

Reply:
> Where/How would you suggest to do that?

I’m assuming that the product data is the same for all integrations (which I didn’t check thoroughly). Then campaignion_tracking could do the comparison and each individual integration could be called only when it actually changes.

Reply:
Ah I see that you already did refactor that. Awesome! :clap: